wasm.md

```diff
@@ -88,14 +88,30 @@ Instructions
 
 ### Sequencing
 
-WebAssembly instructions are space-separated lists of instructions.
+WebAssembly code consists of sequences of declarations (`Decls`).
+We define 4 types of declarations:
+  - Instruction (`Instr`): Administrative or computational instructions. 
+  - Allocation  (`Alloc`): The declarations of `type`, `func`, `table`, `mem` etc.
```

Comment:
The idea is excellent. But I think the names can be improved. This is a major entry point into this semantics file, and hopefully many people are going to use it, read it, debug it, study it, etc., so we should make sure we have the most helpful names possible. I think we can do better, and here's a short essay on the topic.

# Alloc

I don't think `Alloc` is a fitting name, although I think I see what you're going for. "Allocating" memories and tables makes sense, "allocating" a function seems a bit off but okay, seeing how we put it into the main store. But "allocating" an export is just weird. I also think the "allocation" really should happen during module instantiation, which we may want to make into a separate step later, as you mention in the next bullet.

It seems the word of choice in the spec is "definition", and sometimes "declaration". For example, section 2.5 states (emphasis mine):
> A module collects *definitions* for types, functions, tables, memories, and globals. In addition, it can *declare* imports and exports and provide initialization logic in the form of data and element segments or a start function.

Later on we find phrases like
>The data component of a module *defines* a vector of data segments that initialize a range of memory

> The start component of a module declares the function index of a start function

> The imports component of a module *defines* a set of imports

In section 6.6 I only see the word "define" being used for these top-level constructs in a module.

Either way, I think `Def` is the right name for what is now `Alloc`, since all the things in a module can be thought to be "defined":
* types
* funcs
* tables
* mems
* globals
* elem
* data
* start
* imports
* exports

We can think of each of these to be "defining" a module, which is to be later "instantiated".

# Decl

I also do not think it makes sense to talk about a KWasm file containing a list of "declarations" (`$PGM:Decls`), so `Decl` is also slightly off. For example, is `(i32.add (i32.const 1) (i32.const 2))` really a declaration? I think that sounds confusing. This is a place where we are deviating from the spec, and hopefully a lot of people will read and use our spec, so I think it matters to get this right.

The only other two options I can think of is
* `Expr`: Makes sense when looking at Wasm as a Lisp-like, S-expression based language. They are not *pure* expressions, since they may have side effects, but expressions nontheless.
* `Bare`: This keyword makes it explicit that this is what we, in KWasm, allow to appear as bare constructs directly in a file.

To complicate things further, we could either keep `Decl` as in the final use for `module`, and make `Decl` a subsort of `Expr`/`Bare`, which means we have an additional sort (additional confusion), but also an explicit production for what we allow as bare expressions; or, we could make `Expr`/`Bare` the construct for making modules, meaning we would just replace `Decl` with `Expr`/`Bare`.

With a change like this, `Decl` is only used for modules, so it may make more sense to name it `ModuleDecl` for readability.

# Suggestions

I favor using `Def` instead of `Alloc`, using `ModuleDecl` for modules, and making `Bare` the top-level construct. So below here, we would have productions like:

```k
    syntax Bare    ::= Instr | Def | ModuleDecl
    syntax Bares  ::= Instrs
                    | Defs
                    | ModuleDecls
                    | List{Bare, ""} [klabel(listBare)]
```

I think that's pretty readable and gives a good overview of what we allow. We should also have a sentence on that we are more permissive than Wasm, in that we do allow instructions and definitions at the top level.

## The other suggestions, in order of preference:

* `Instr`, `Def`, `Bare` 
* `Instr`, `Def`, `ModuleDecl`, `Expr` 
* `Instr`, `Def`, `Expr`

Reply:
New suggestion:

```k
    syntax Stmt   ::= Instr | Defn | ModuleDecl
    syntax Stmts  ::= Instrs
                    | Defs
                    | ModuleDecls
                    | List{Bare, ""} [klabel(listBare)]
```

Reply:
So maybe we can try:

```
    syntax Stmt   ::= Instr | Defn | ModuleDecl
    syntax Instrs  ::= List{Instr, ""} [klabel(listInstr)]
    syntax Stmts  ::= List{Bare, ""} [klabel(listStmt)]

rule <k> IS:Instrs => asStatements(IS) ... </k>

syntax Stmts ::= asStatements ( Instrs ) [function]
rule asStatements(.Instrs) => .Stmts
rule asStatements(listInstr(I, IS)) => listStmt(I, asStatements(IS))
```

Reply:
Might have to say `.listInstr` instead `.Instrs`.

Reply:
And no `ModuleDecl`, right? Just let `(module ...)` be of sort `Stmt`.